### ISSUE TITLE
[mono] UnsafeAccessor to a virtual method on a base class through a derived type asserts

### ISSUE THREAD

**Issue by vitek-karas** [dotnet/runtime#89212]
```C#
    class InheritanceBase
    {
        protected virtual string BaseVirtual() => $"{nameof(InheritanceBase)}.{nameof(BaseVirtual)}";
    }

    class InheritanceDerived : InheritanceBase
    {
    }

    [Fact]
    public static void Verify_InheritanceMethodResolution()
    {
        var instance = new InheritanceDerived();
        Assert.Equal($"{nameof(InheritanceBase)}.{nameof(BaseVirtual)}", BaseVirtual(instance));

        [UnsafeAccessor(UnsafeAccessorKind.Method, Name = nameof(BaseVirtual))]
        extern static string BaseVirtual(InheritanceDerived target);
    }
```

This fails with
```
* Assertion at F:\dotnet\runtime4\src\mono\mono\metadata\marshal-lightweight.c:2508, condition `target_method->klass == target_class' not met
```

@lambdageek - could you please route this to the right area/person

**Comment by vitek-karas** [comment]
Actually this fails on NativeAOT as well (at runtime!):

```
System.MissingMethodException: Method not found: 'BaseVirtual'.
   at Internal.Runtime.TypeLoaderExceptionHelper.CreateMissingMethodException(ExceptionStringID, String) + 0x63
   at Internal.Runtime.CompilerHelpers.ThrowHelpers.ThrowMissingMethodException(ExceptionStringID, String) + 0x25
   at UnsafeAccessorsTests.<Verify_InheritanceMethodResolution_BaseViaDerived>g__BaseVirtual|33_0(UnsafeAccessorsTests.InheritanceDerived target) + 0x1f
   at UnsafeAccessorsTests.Verify_InheritanceMethodResolution_BaseViaDerived() + 0x4b
```

CoreCLR works and the same code without `UnsafeAccessor`:
```C#
var instance = new InheritanceDerived();
instance.BaseVirtual();
```
will also work (probably on all runtimes, although I didn't try that).

@AaronRobinsonMSFT @MichalStrehovsky what do you think - should this work?

**Comment by MichalStrehovsky** [comment]
Well method `BaseVirtual` is not on `InheritanceDerived` so this is not expected to work.

This is what I brough up in https://github.com/dotnet/runtime/pull/86932#discussion_r1218888771 and was rejected.

**Comment by AaronRobinsonMSFT** [comment]
@vitek-karas If I update as following it works, right?

```csharp
        [UnsafeAccessor(UnsafeAccessorKind.Method, Name = nameof(BaseVirtual))]
        extern static string BaseVirtual(InheritanceBase target);
```

If so, I don't see why one would declare this as `InheritanceDerived`, when the target method is defined on a different type. It isn't as if I can know about non-public methods by looking at a derived type. The first thing I would do would be to look at `InheritanceDerived`, not see a non-public method and be rather confused.

**Comment by MichalStrehovsky** [comment]
Does this work with CoreCLR?

**Comment by vitek-karas** [comment]
Yes - it does work on CoreCLR.

**Comment by MichalStrehovsky** [comment]
> Yes - it does work on CoreCLR.

@AaronRobinsonMSFT I think this should be in 8.0. We need to either block this in CoreCLR, or fix in NativeAOT/Mono.

**Comment by MichalStrehovsky** [comment]
Btw, I think the CoreCLR behavior is the correct one

**Comment by vitek-karas** [comment]
Personally I agree with Michal's comment in the other issue that this "should" work. That said I'd be fine if we said that we don't want this to work. But then all runtimes should behave consistently.
Especially differences between CoreCLR and NativeAOT are dangerous (since we promise that AOTed apps behave the same as non-AOTed app)

**Comment by AaronRobinsonMSFT** [comment]
> Personally I agree with Michal's comment in the other issue that this "should" work. 

What needs to happen then in NativeAOT? Based on some offline conversations with @fanyang-mono, I believe she is looking into enabling it on mono.

**Comment by MichalStrehovsky** [comment]
> What needs to happen then in NativeAOT?

We'd walk into the base type, like we do when resolving a MemberRef:

https://github.com/dotnet/runtime/blob/9343b2f61cc9e78be2c0d607bb5db0cc588aec49/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaModule.cs#L525

(I assume it matches what CoreCLR does in the `UnsafeAccessor` case too.)

**Comment by AaronRobinsonMSFT** [comment]
> (I assume it matches what CoreCLR does in the UnsafeAccessor case too.)

Yep. I just used the `MethodIterator`. Seems like it does that under the covers but it wasn't obvious.

**Comment by lambdageek** [comment]
I think the work in Mono is just to replace the assert that the target class is equal to the method's class by a check that the target class is assignable to the method's class.

**Comment by fanyang-mono** [comment]
This has been resolved via #89217.
